cmd/google-cloud-dns-healthcheck/run.go

```diff
@@ -0,0 +1,173 @@
+	RecordName       string   `short:"n" long:"record-name" required:"true" env:"RECORD_NAME" description:"Dns record name"`
+	Project          string   `short:"p" long:"project" required:"true" env:"PROJECT" description:"Google project"`
+	ManagedZone      string   `short:"z" long:"managed-zone" required:"true" env:"MANAGED_ZONE" description:"Google DNS managed zone"`
+	HealthcheckPath  string   `short:"c" long:"healthcheck-path" required:"true" env:"HEALTHCHECK_PATH" description:"HealthcheckPath"`
+	Rrdatas          []string `short:"r" long:"rrdatas" required:"true" env:"RRDATAS" env-delim:"," description:"Expected rrdatas (in comma-separated format from env variable)"`
+	HttpTimeout      int64    `short:"t" long:"http-timeout" env:"HTTP_TIMEOUT" default:"5" description:"Expected rrdatas comma-separated format"`
+	HttpScheme       string   `short:"s" long:"http-scheme" env:"HTTP_SCHEME" default:"http" description:"Http scheme"`
+	DryRun           bool     `short:"d" long:"dry-run" description:"Run without performing any modification"`
+}
+
+func (r *RunCommand) ExecuteContext(ctx context.Context, args []string) error {
+	dnsService, err := r.getDnsService(ctx)
+	if err != nil {
+		log.Errorf(err, "Error getting google dns service handler")
+		return err
+	}
+
+	record, err := r.getDnsRecord(dnsService)
+	if err != nil {
+		log.Errorf(err, "Error getting dns record")
+		return err
+	}
+
+	if len(r.intersection(record.Rrdatas)) == 0 {
+		err := fmt.Errorf("Wrong rrdatas value")
+		log.Errorf(err, fmt.Sprintf("No intersection between given rrdatas %v and record's rrdatas %v", r.Rrdatas, record.Rrdatas))
+		return err
+	}
+
+	healthyRrdatas := r.checkRrdatas()
+
+	switch len(healthyRrdatas) {
+	case len(record.Rrdatas):
+		log.Debugf("All rrdatas are healthy")
+		return nil
+	case 0:
+		log.Warningf("All rrdatas are unhealthy. We won't touch the record")
+		return nil
+	default:
+		log.Debugf("Updating record")
+		if err := r.updateDnsRecord(dnsService, record, healthyRrdatas); err != nil {
+			log.Errorf(err, "Error changing DNS RecordSet")
+			return err
+		}
+		return nil
+	}
+
+}
+
+func (r *RunCommand) getDnsService(ctx context.Context) (*dns.Service, error) {
+	c, err := google.DefaultClient(ctx, dns.CloudPlatformScope)
+	if err != nil {
+		return nil, err
+	}
+
+	dnsService, err := dns.New(c)
+	if err != nil {
+		return nil, err
+	}
+
+	return dnsService, nil
+}
+
+func (r *RunCommand) getDnsRecord(dnsService *dns.Service) (*dns.ResourceRecordSet, error) {
+	resp, err := dnsService.ResourceRecordSets.List(
+		r.Project, r.ManagedZone).Name(r.RecordName).Type(ResourceRecordSetTypeA).Do()
+
+	if err != nil {
+		return nil, fmt.Errorf("Record not found: %s", r.RecordName)
+	}
+
+	if len(resp.Rrsets) > 1 {
+		return nil, fmt.Errorf("Only expected 1 record set, got %d", len(resp.Rrsets))
+	}
+
+	return resp.Rrsets[0], nil
```

Comment:
More a question than a comment: will it always error on none found? Otherwise `[0]` might be nil

Reply:
Yes, that's handled in the above conditional